### PR TITLE
Port FullFx fix for SocketConnection sync/async timeout ignore problem.

### PR DESCRIPTION
**Pull request by StephenBonikowsky**
* [Changeset 1615113](http://vstfdevdiv.redmond.corp.microsoft.com:8080/DevDiv2/DevDiv/_versionControl/changeset/1615113?_a=compare&path=%24%2FDevDiv%2FFX%2FFeature%2FNetFXDev1%2FNDP%2Fcdf%2Fsrc%2FWCF%2FServiceModel%2FSystem%2FServiceModel%2FChannels%2FSocketConnection.cs) in VSTS.

### CODE REVIEW

**Comment by StephenBonikowsky** [comment]
Going into master first then will cherry-pick to release/3.1.0.

@dasetser There is a fair amount of difference between the desktop and core versions of SocketConnection.cs so I had to port the changes manually and not all code paths were present.
Please review.

**Comment by mconnew** [comment]
@StephenBonikowsky, the reason they are different is because I had to modify the implementation because of two reasons.  
1. Socket.Close(int timeout) wasn't originally available in .NET Core so I had to rewrite the close code to use socket lingering capabilities.
2. Neither System.Net.Sockets or SslStream wasn't available for UWP so I had to rewrite everything on top of the UWP native classes for sockets which didn't have a 1:1 mapping of functionality so I refactored everything to a base class and the concrete implementation.

Neither of this problems exist today so my preference would be to replace the entire SocketConnection class with the implementation from .NET Framework as there are some compromises with the current implementation, e.g. close timeout only works with a granularity of 1 second.

**Comment by dasetser** [comment]
@StephenBonikowsky It looks like you ported the fix correctly to the current implementation. However, I agree with Matt that it would be better if we could port the entire file over again if we can get rid of the compromises in the current implementation.

**Comment by StephenBonikowsky** [comment]
Sounds good, I'll find Matt's commits and see if reverting them makes sense and then porting the SocketConnection class as it is in the full framework.

**Comment by StephenBonikowsky** [comment]
This was not a straight forward port, no copy and paste and just fix formatting issues. There are quite a few things not present in Core that I had to account for. Particularly related to tracing.

Please review this carefully.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs [hunk 1]
```diff
@@ -316,33 +340,68 @@ private void CloseAsyncAndLinger()
 
             try
             {
-                // A FIN (shutdown) packet has already been sent to the remote host and we're waiting for the remote
-                // host to send a FIN back. A pending read on a socket will complete returning zero bytes when a FIN
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs]: This comment is worth keeping as it describes what the code is doing here.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs [hunk 2]
```diff
@@ -486,7 +522,7 @@ private void ThrowIfNotOpen()
             {
                 throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
                     ConvertObjectDisposedException(new ObjectDisposedException(
-                    GetType().ToString(), SR.SocketConnectionDisposed), TransferOperation.Undefined));
+                    GetType().ToString(), SR.Format(SR.SocketConnectionDisposed)), TransferOperation.Undefined));
             }
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs]: SR.Format isn't needed when there are no parameters for the string resource

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs [hunk 3]
```diff
@@ -496,37 +532,77 @@ private void ThrowIfClosed()
...
             if (_timeoutErrorString != null)
             {
-                return ConvertTimeoutErrorException(originalException, _timeoutErrorString, _timeoutErrorTransferOperation);
+                return ConvertTimeoutErrorException(originalException, transferOperation, _timeoutErrorString, _timeoutErrorTransferOperation);
             }
             else if (_aborted)
             {
-                return new CommunicationObjectAbortedException(SR.SocketConnectionDisposed, originalException);
+                return new CommunicationObjectAbortedException(SR.Format(SR.SocketConnectionDisposed), originalException);
             }
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs]: Remove SR.Format as unnecessary, just use the resource string directly

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs [hunk 4]
```diff
@@ -535,42 +611,42 @@ private Exception ConvertObjectDisposedException(ObjectDisposedException origina
...
-                (int)socketException.SocketErrorCode == UnsafeNativeMethods.WSAECONNABORTED ||
-                (int)socketException.SocketErrorCode == UnsafeNativeMethods.WSAECONNRESET)
+            if (socketException.ErrorCode == UnsafeNativeMethods.WSAENETRESET ||
+                socketException.ErrorCode == UnsafeNativeMethods.WSAECONNABORTED ||
+                socketException.ErrorCode == UnsafeNativeMethods.WSAECONNRESET)
             {
                 if (aborted)
                 {
-                    return new CommunicationObjectAbortedException(SR.TcpLocalConnectionAborted, originalException);
+                    return new CommunicationObjectAbortedException(SR.Format(SR.TcpLocalConnectionAborted), originalException);
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs]: Remove SR.Format

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs [hunk 5]
```diff
@@ -662,12 +724,12 @@ private void TraceWriteStart(int size, bool async)
...
                 throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
-                    ConvertSendException(socketException, TimeSpan.MaxValue));
+                    ConvertSendException(socketException, TimeSpan.MaxValue, _asyncSendTimeout));
             }
             catch (ObjectDisposedException objectDisposedException)
             {
                 Exception exceptionToThrow = ConvertObjectDisposedException(objectDisposedException, TransferOperation.Write);
-                if (ReferenceEquals(exceptionToThrow, objectDisposedException))
+                if (object.ReferenceEquals(exceptionToThrow, objectDisposedException))
                 {
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs]: I think there should be a VS Roslyn suggested fix here to simplify the code to remove object. as it's unnecessary. If there is, choose to apply the fix to the entire document.

**Comment by mconnew** [comment]
Just a couple of small issues around string resources, otherwise LGTM. Fix those SR.Format uses and we can call it good!

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs (hunk 1 above)
Comment (StephenBonikowsky) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs]: 👍

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs (hunk 2 above)
Comment (StephenBonikowsky) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs]: 👍

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs (hunk 3 above)
Comment (StephenBonikowsky) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs]: 👍

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs (hunk 4 above)
Comment (StephenBonikowsky) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs]: 👍

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs (hunk 5 above)
Comment (StephenBonikowsky) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs]: 👍

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs [hunk 6]
```diff
@@ -46,37 +56,100 @@ internal class SocketConnection : IConnection
...
-            AsyncReadBuffer = _connectionBufferPool.Take();
-            AsyncReadBufferSize = AsyncReadBuffer.Length;
-            _sendTimeout = _receiveTimeout = TimeSpan.MaxValue;
-            _closeState = CloseState.Open;
-            _socket.SendBufferSize = _socket.ReceiveBufferSize = AsyncReadBufferSize;
-            _sendTimeout = _receiveTimeout = TimeSpan.MaxValue;
-        }
+            _readBuffer = connectionBufferPool.Take();
+            _asyncReadBufferSize = _readBuffer.Length;
+            socket.SendBufferSize = socket.ReceiveBufferSize = _asyncReadBufferSize;
```
[the remark this answers is not in the document]
Comment (StephenBonikowsky) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SocketConnection.cs]: 👍

**Comment by StephenBonikowsky** [comment]
The new tests added with #4000 are failing because the test server hasn't been updated. Will update it manually and then re-run the checks.

**Comment by mconnew** [comment]
:shipit:

**Pull request merged**